Added migration to fix incorrect mrr events

rshbhgrg opened the pull request:
refs https://github.com/TryGhost/Team/issues/595

Due to a bug in mrr_delta calculation, we ended up reducing the MRR delta by twice the original amount when a subscription goes from active to canceled and storing it in `members_paid_subscription_events` table, which is used to show the MRR chart on Dashboard. The way we identify the incorrect events in the table which got the double negative value is by checking if they match certain criteria - Both `from_plan` and `to_plan` have same value as this is the only case this bug always happens in since a subscription changes status on same plan.

This migration halves the `mrr_delta` for incorrect events to restore the correct MRR change for the site.

rshbhgrg committed:
Added migration to fix incorrect mrr events

refs https://github.com/TryGhost/Team/issues/595

Due to a bug in mrr_delta calculation, we ended up reducing the MRR delta by twice the original amount when a subscription goes from active to canceled. The way we find the incorrect events which got the double negative value is by checking

- Both `from_plan` and `to_plan` have same value as this happens when subscription is canceled
- The MRR_DELTA value is negative, as only those events which had cancelation will be effected
- The MRR_DELTA is even number as it will be twice of the original subscription amount

This migration halves the MRR_DELTA for incorrect events to restore the correct MRR change for the site.

rishabhgrg commented on core/server/data/migrations/versions/4.2/01-fix-incorrect-mrr-delta-events.js:
```diff
@@ -0,0 +1,34 @@
+        }
+    },
+    async function down() {
+        // noop
```
@allouis I was wondering if a noop is correct in this case or we should look at reverting the above calculation for the `down` migration here 🤔 IMO noop is better because we don't want to (un)fix the mrr calculation, but not feeling 100% about the right thing

allouis replied:
We need a `down` migration for sure, because otherwise a rollback will mean that running the migration again will break the values!

allouis commented:
Is it possible for a row to meet the first criteria (from_plan is the same as to_plan) and it not meet the second two?

rishabhgrg commented:
@allouis 🤔 We probably could, its unlikely there will be cases which match the first and not the other 2, the other 2 are more of a safety net. Though yeah the UPDATE query will much better if we just consider the first

rshbhgrg committed:
Cleaned up to check for plan equality

The other 2 checks for negative and even values are redundant in this case as we will only have this incorrect mrr scenario when both plans are same and only the status has changed to `canceled` . For other status change events the mrr delta will anyways be 0.

rishabhgrg commented:
@allouis Had another 🤔 session and updated the migration to directly update based on `from_plan` = `to_plan` criteria as the others are redundant

daniellockyer commented:
@rishabhgrg I've noticed the following:

```
> select mrr_delta from members_paid_subscription_events where from_plan=to_plan limit 5;
+-----------+
| mrr_delta |
+-----------+
|     -1000 |
|     -1000 |
|     -1000 |
|      -916 |
|     -1000 |
+-----------+
5 rows in set (0.00 sec)

> select mrr_delta/2 from members_paid_subscription_events where from_plan=to_plan limit 5;
+-------------+
| mrr_delta/2 |
+-------------+
|   -500.0000 |
|   -500.0000 |
|   -500.0000 |
|   -458.0000 |
|   -500.0000 |
+-------------+
5 rows in set (0.00 sec)

> select round(mrr_delta/2) from members_paid_subscription_events where from_plan=to_plan limit 5;
+--------------------+
| round(mrr_delta/2) |
+--------------------+
|               -500 |
|               -500 |
|               -500 |
|               -458 |
|               -500 |
+--------------------+
```

Unless we do a `round`, we're storing the values as floats. Does this matter? 🤔

rishabhgrg commented:
@daniellockyer 🤔 I don't think it matters because it seems when the UPDATE sets it back its converted to INTEGER as its working properly. But, its probably on the safer side to introduce `ROUND` to the SET so i'll update that

rshbhgrg committed:
Rounded MRR value on set

allouis reviewed (changes requested):
Looks good, just missing the `down` migration :)

rshbhgrg committed:
Added down migration to reverse the delta change

tpatel reviewed (commented):
The code looks great, but I still have to run it locally 😬 doing my best to test it quickly

tpatel reviewed (approved):
Looks good to me, well done 🙌
